airflow/plugins/utils.py

```diff
@@ -81,7 +153,369 @@ def get_successfully_downloaded_feeds(execution_date):
     return status[lambda d: d.status == "success"]
 
 
-def make_name_bq_safe(name: str):
-    """Replace non-word characters.
-    See: https://cloud.google.com/bigquery/docs/reference/standard-sql/lexical#identifiers."""
-    return str.lower(re.sub("[^\w]", "_", name))  # noqa: W605
+# class GTFSFeedType(str, Enum):
+#     schedule = "schedule"
+#     rt = "rt"
+
+# TODO: consider moving this to calitp-py or some other more shared location
+class GTFSFeedType(str, Enum):
+    schedule = "schedule"
+    service_alerts = "service_alerts"
+    trip_updates = "trip_updates"
+    vehicle_positions = "vehicle_positions"
+
+
+class AirtableGTFSDataRecord(BaseModel):
+    name: str
+    uri: Optional[str]
+    data: GTFSFeedType
+    schedule_to_use_for_rt_validation: Optional[List[str]]
+    auth_query_param: Dict[str, str] = {}
+
+    class Config:
+        extra = "allow"
+
+    @validator("data", pre=True, allow_reuse=True)
+    def convert_feed_type(cls, v):
+        if "schedule" in v.lower():
+            return GTFSFeedType.schedule
+        elif "vehicle" in v.lower():
+            return GTFSFeedType.vehicle_positions
+        elif "trip" in v.lower():
+            return GTFSFeedType.trip_updates
+        elif "alerts" in v.lower():
+            return GTFSFeedType.service_alerts
+        return v
+
+    @property
+    def schedule_url(self) -> Optional[AnyUrl]:
+        # TODO: implement me
+        raise NotImplementedError
+
+    @property
+    def auth_header(self) -> Dict[str, str]:
+        return {}
+
+    @property
+    def base64_encoded_url(self) -> str:
+        # see: https://docs.python.org/3/library/base64.html#base64.urlsafe_b64encode
+        # we care about replacing slashes for GCS object names
+        # can use: https://www.base64url.com/ to test encoding/decoding
+        # convert in bigquery: https://cloud.google.com/bigquery/docs/reference/standard-sql/string_functions#from_base64
+        return base64.urlsafe_b64encode(self.uri.encode()).decode()
+
+    def build_request(self, auth_dict: dict) -> Request:
+        filled_auth_params = {k: auth_dict[v] for k, v in self.auth_query_param.items()}
+        filled_auth_headers = {k: auth_dict[v] for k, v in self.auth_header.items()}
+        # inspired by: https://stackoverflow.com/questions/18869074/create-url-without-request-execution
+        return Request(
+            "GET", url=self.uri, params=filled_auth_params, headers=filled_auth_headers
+        )
+
+
+class PartitionedGCSArtifact(BaseModel, abc.ABC):
+    """
+    This class is designed to be subclassed to model "extracts", i.e. a particular
+    download of a given data source.
+    """
+
+    filename: str
+
+    class Config:
+        json_encoders = {
```

Comment:
do you still need this here with the `PETITION_SERIALIZERS` above? seems the same thing?

Reply:
ah yes we do, the partition de/serializers are specific to partitions, whereas `json_encoders` configure Pydantic-serialized JSON